geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java

```diff
@@ -5311,10 +5312,20 @@ public void invokeRegionEntrySynchronizationListenersAfterSynchronization(
   }
 
   @Override
-  public Object convertPdxInstanceIfNeeded(Object obj) {
+  public Object convertPdxInstanceIfNeeded(Object obj, boolean preferCD) {
     Object result = obj;
-    if (!this.getPdxReadSerialized() && obj instanceof PdxInstance) {
-      result = ((PdxInstance) obj).getObject();
+    if (obj instanceof InternalPdxInstance) {
+      InternalPdxInstance pdxInstance = (InternalPdxInstance) obj;
+      if (preferCD) {
+        try {
+          result = new PreferBytesCachedDeserializable(pdxInstance.toBytes());
+        } catch (IOException ignore) {
+          // Could not convert pdx to bytes here; it will be tried again later
+          // and an exception will be thrown there.
+        }
+      } else if (!this.getPdxReadSerialized()) {
+        result = ((PdxInstance) obj).getObject();
```

Comment:
We could call pdxInstance.getObject without cast; instead obj.getObject.

Reply:
I removed this cast